plugins/cilium-cni/cilium-cni.go

```diff
@@ -389,25 +390,74 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 
 	conf := *configResult.Status
 
-	podName := string(cniArgs.K8S_POD_NAMESPACE) + "/" + string(cniArgs.K8S_POD_NAME)
-	ipam, err = c.IPAMAllocate("", podName, true)
-	if err != nil {
-		err = fmt.Errorf("unable to allocate IP via local cilium agent: %s", err)
-		return
-	}
+	if conf.IpamMode == ipamOption.IPAMDelegatedPlugin {
+		// Skip the call to cilium-agent IPAMAllocate and use the delegated IPAM plugin instead.
+		var ipamRawResult cniTypes.Result
+		var ipamResult *cniTypesVer.Result
+		ipamRawResult, err = cniInvoke.DelegateAdd(context.TODO(), n.IPAM.Type, args.StdinData, nil)
+		if err != nil {
+			err = fmt.Errorf("delegated IPAM cmd ADD failed: %s", err)
+			return
+		}
 
-	if ipam.Address == nil {
-		err = fmt.Errorf("Invalid IPAM response, missing addressing")
-		return
-	}
+		// CNI spec says if an error occurs, we should call the delegated IPAM plugin
+		// again with a DEL command so it can release the IPs.
+		defer func() {
+			if err != nil {
+				cniInvoke.DelegateDel(context.TODO(), n.IPAM.Type, args.StdinData, nil)
```

Comment:
I've left the `context.TODO()`s for now because I didn't want to make changes that could impact non-delegated-plugin codepaths. Interestingly, the bridge reference plugin also uses `context.TODO`, which makes me think it might be relying on the container runtime to kill the CNI process after some timeout.